File: node/fenwicktree/test/types/JSFenwickTree.test.js
const { describe, test, expect, beforeEach } = require('@jest/globals');
const JSFenwickTree = require('../../src/types/JSFenwickTree');

describe('Fenwick tree test', () => {

    let mTree;
    let mArray;

    beforeEach(() => {
        mTree = null;
        mArray = null;
    });

    test('Test sum', () => check() );

    test('the range method', () => {
        expect(tree().range(3, 5)).toBe(12);
    });

    test('test update', () => {
        tree().update(3, 10);
        testArray()[2] = 10;
        check();
    });

    //*******************
    // Private Members **
    //*******************

    const check = () => {
        for(let i = 1; i <= testArray().length; i++) {
            expect(sum(i)).toBe(tree().sum(i));
        }
    }

    const sum = (index) => testArray().reduce((total, next, i) => total += i > index - 1 ? 0 : next, 0);

    const tree = () => mTree === null ? mTree = new JSFenwickTree(testArray()) : mTree;

    const testArray = () => mArray === null ? mArray = [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 ] : mArray;
});

File: java/fenwicktree/src/test/java/com/js/datastructuresalgorithms/fenwicktree/JSFenwickTreeTest.java
package com.js.datastructuresalgorithms.fenwicktree;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class JSFenwickTreeTest extends TestBase<JSFenwickTree> {

    private final int[] mArray = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};

    @Test
    public void sum()
    {
        check();
    }

    @Test
    public void range()
    {
        Assertions.assertEquals(sum(6) - sum(3), testCase().range(4, 6));
    }

    @Test
    public void update()
    {
        mArray[2] = 10;
        testCase().update(2, 10);
        check();
    }

    protected JSFenwickTree createTest()
    {
        return new JSFenwickTree(mArray);
    }

    //*******************
    // Private Members **
    //*******************

    private void check()
    {
        for(int i = 1; i <= mArray.length; i++) {
            Assertions.assertEquals(sum(i), testCase().sum(i));
        }
    }

    private int sum(int to) {
        int total = 0;
        for(int i = 0; i < to; i++) {
            total += mArray[i];
        }
        return total;
    }
}